File: Έγγραφα SEM4105/Λύσεις ασκήσεων/08. Όγδοο εργαστήριο/3/deriv.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>


// max abs value of a(row,:), b(row)
double maxval(int n, double const a[], double const b[], int row)
{
  int mj = 0;
    
  for (int j = 1; j < n; ++j) {
    if (fabs(a[row+j*n]) > fabs(a[row+mj*n])) {
      mj=j;
    }
  }

  double g = fmax(fabs(b[row]), fabs(a[row+mj*n]));
    
  return g;
}

void swap(double * a, double *b)
{
  double t = *a;
  *a = *b;
  *b = t;
}

void pivot(int k, int n, double a[], double b[])
{
  int mi = k;
  double mvmi = maxval(n,a,b,mi);
  for (int i = k+1; i < n; ++i) {
    double  mvi = maxval(n,a,b,i);
    
    if (fabs(a[i+k*n] / mvi) > fabs(a[mi+k*n] / mvmi) ) {
      mi = i;
      mvmi = mvi;
    }
  }
  
  for (int j=k; j < n; ++j) {
    swap(&a[mi+j*n], &a[k+j*n]);
  }
  swap(&b[mi], &b[k]);
  
}



void triang(int n, double a[], double  b[])
{
  for (int k = 0; k < n-1; ++k) {

    pivot(k,n,a,b);
	
    for (int i = k+1; i < n; ++i) {
      double const g = -a[i+n*k] / a[k+n*k];
      for (int j = k; j < n; ++j) {
	a[i+j*n] += g * a[k +j*n];
      }
      b[i] += g * b[k];
    }
  }
}


void backsub(int n, double const a[], double const b[], double x[])
{
  for (int i = n-1; i >= 0; --i) {
    double s = b[i];
    for (int j = i+1; j < n; ++j) {
      s -= a[i+j*n] * x[j];
    }
    x[i] = s / a[i+i*n];
  }	
}


void gauss(int n, double a[], double b[], double x[])
{
  triang(n, a, b);
  backsub(n,a,b,x);
}



/*     Read data file */
void read_data(char const * fname, double **x, double **y, int * nptr)
{
  FILE *in = fopen(fname, "r");

  fscanf(in,"%d\n", nptr);

  *x = malloc(*nptr * sizeof(double));
  *y = malloc(*nptr * sizeof(double));
  
  for (int i = 0; i < *nptr; ++i) {
    double a, b;
    fscanf(in,"%lf%lf", &a, &b);
    (*x)[i] = a;
    (*y)[i] = b;
  }
  fclose(in);
}



/*  f^{(order)}_i (x) */
double f(int i, int order, double x)
{
  double v;
    
  switch (order) {
  case 0:    /* => function value */
    v = pow(x,i);
    break;
	
  case 1:    /* => first derivative */
    if (i == 0) {
      v = 0.0;
    } else {
      v= i * pow(x,i-1);
    }
    break;

  case 2: /* => second derivative */
    if (i == 0 || i == 1) {
      v = 0.0;
    } else {
      v= i*(i-1)*pow(x,i-2);
    }
    break;
  default:
    fprintf(stderr, "%s", "not implemented\n");
    exit(-1);
      
  }

  return v;
}


void sys(int n, double const x[], double xbar, int order, double a[],
	 double b[])
{
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < n; ++j) {
      a[i+j*n] = f(i, 0, x[j]);
    }
    b[i] = f(i,order,xbar);
  }

}

double correct(int order, double x)
{
  switch (order) {
  case 0:
    return pow(sin(x),2);
  case 1:
    return 2*sin(x)*cos(x);
  case 2:
    return 2*cos(2*x);
  default:
    fprintf(stderr, "%s", "not implemented\n");
    exit(-1);
  }

}

int main(void)
{
  int n;
  double *x, *y;
  read_data("points2.dat", &x, &y, &n);

  double *a = malloc(n*n*sizeof(double));
  double *b = malloc(n*sizeof(double));
  double *sol = malloc(n*sizeof(double));

  double xbar = 2.5;
    
  for (int order = 1; order <= 2; ++order) {
    sys(n, x, xbar, order, a, b);
	
    gauss(n,a,b,sol);
	
    double s = 0.0;
    for (int i = 0; i < n; ++i) {
      s += sol[i] * y[i];
    }
    
    free(x);
    free(y);
    free(sol);
    free(b);
    free(a);
    
    printf("%d derivative at %g is %g\n", order, xbar, s);
    printf("difference from correct value %g\n", s - correct(order,xbar));
  }

  return 0;
}

File: Έγγραφα SEM4105/Λύσεις ασκήσεων/08. Όγδοο εργαστήριο/3/deriv.f90
PROGRAM derivative
  IMPLICIT NONE
  DOUBLE PRECISION, ALLOCATABLE :: xp(:), yp(:)
  DOUBLE PRECISION :: x, der1, der2

  INTERFACE
     SUBROUTINE read_data(fname,  x, y)
       IMPLICIT NONE
       CHARACTER(*), INTENT (in) :: fname
       DOUBLE PRECISION, ALLOCATABLE, INTENT (out) :: x(:), y(:)
     END SUBROUTINE read_data

     FUNCTION deriv(order, xp, yp, x)
       IMPLICIT NONE
       INTEGER,          INTENT (in) :: order
       DOUBLE PRECISION, INTENT (in) :: xp(:), yp(:)
       DOUBLE PRECISION, INTENT (in) :: x
       DOUBLE PRECISION :: deriv
     END FUNCTION deriv

  END INTERFACE

  CALL read_data("points.dat", xp, yp)

  PRINT *, "Δώσε το σημείο υπολογισμού των παραγώγων: "
  READ *, x

  der1 = deriv(1, xp, yp, x)
  der2 = deriv(2, xp, yp, x)

  PRINT *, "Η πρώτη παράγωγος είναι", der1
  PRINT *, "Η διαφορά από την ακριβή τιμή είναι ", der1 - SIN(2d0*x)
  PRINT *
  PRINT *, "Η δεύτερη παράγωγος είναι", der2
  PRINT *, "Η διαφορά από την ακριβή τιμή είναι ", der2 - 2d0*COS(2d0*x)
END PROGRAM derivative


FUNCTION deriv(order, xp, yp, x)

  IMPLICIT NONE
  INTEGER,          INTENT (in) :: order
  DOUBLE PRECISION, INTENT (in) :: xp(:), yp(:)
  DOUBLE PRECISION, INTENT (in) :: x
  DOUBLE PRECISION :: deriv

  DOUBLE PRECISION, ALLOCATABLE :: A(:,:), b(:), solution(:)
  INTEGER :: n, k

  INTERFACE
     SUBROUTINE TRIANG(A, B)
       IMPLICIT NONE
       DOUBLE PRECISION, INTENT (inout) :: A(:,:), B(:)
     END SUBROUTINE TRIANG

     SUBROUTINE BACKSU(A, B, X)
       IMPLICIT NONE
       DOUBLE PRECISION, INTENT (in)  :: A(:,:), B(:)
       DOUBLE PRECISION, INTENT (out) :: X(:)
     END SUBROUTINE BACKSU
  END INTERFACE

  n = SIZE(xp)

  ALLOCATE(A(n,n), b(n))

  SELECT CASE (order)
  CASE (1)
     b(1) = 0.0d0   ! πρώτη παράγωγος του 1
     b(2) = 1.0d0   ! πρώτη παράγωγος του x     
     DO k=3,n
        b(k)   = (k-1) * x**(k-2)  ! πρώτη παράγωγος του x**(k-1)
     ENDDO
  CASE (2) 
     b(1) = 0.0d0   ! δεύτερη παράγωγος του 1
     b(2) = 0.0d0   ! δεύτερη παράγωγος του x     
     b(3) = 2.0d0   ! δεύτερη παράγωγος του x^2

     DO k=4,n
        b(k) = (k-1) * (k-2) * x**(k-3)  ! δεύτερη παράγωγος του x**(k-1)
     ENDDO
  CASE default
     STOP 'not implemented'
  END SELECT

  a(1,:) = 1.0d0
  DO k=2,n
     a(k,:) = xp**(k-1)
  ENDDO

  CALL TRIANG(A, B)

  ALLOCATE (solution(n))
  CALL BACKSU(A, B, solution)

  deriv = DOT_PRODUCT(solution, yp)

END FUNCTION deriv


!     Read data file
SUBROUTINE read_data(fname, x, y)
  IMPLICIT NONE
  CHARACTER(*), INTENT (in) :: fname
  DOUBLE PRECISION, ALLOCATABLE, INTENT (out) :: x(:), y(:)

  INTEGER, PARAMETER ::  UNITNO = 54
  INTEGER :: n, i

  OPEN (UNIT = UNITNO, FILE = fname, status = 'old', action = 'read')
  READ (UNITNO, *) n

  ALLOCATE(X(n), Y(n))

  DO I=1,n
     READ (UNITNO, *) X(I), Y(I)
  ENDDO

  CLOSE (UNITNO)
END SUBROUTINE read_data


!!!! Gauss 


!     MAKE LOWER TRIANGULAR
SUBROUTINE TRIANG(A, B)
  IMPLICIT NONE
  DOUBLE PRECISION, INTENT (inout) :: A(:,:), B(:)

  INTERFACE
     SUBROUTINE pivot(a, b, i)
       IMPLICIT NONE

       DOUBLE PRECISION, INTENT (inout) :: A(:,:), B(:)
       INTEGER, INTENT (in) :: i
     END SUBROUTINE pivot
  END INTERFACE

  INTEGER :: N, I, K
  DOUBLE PRECISION :: G

  N = SIZE(B)

  DO K = 1, N
     CALL pivot(a, b, k)
     DO I=K+1, N
        G = -A(I,K) / A(K,K)
        A(I,K:N) = A(I,K:N) + A(K,K:N) * G
        B(I) = B(I) + B(K) * G
     ENDDO
  ENDDO

END SUBROUTINE TRIANG


!     BACKSUBSTITUTION
SUBROUTINE BACKSU(A, B, X)
  IMPLICIT NONE
  DOUBLE PRECISION, INTENT (in)  :: A(:,:), B(:)
  DOUBLE PRECISION, INTENT (out) :: X(:)

  INTEGER :: K, N

  N = SIZE(b)

  DO K = N,1,-1
     X(K) = (B(K) - DOT_PRODUCT(A(K,K+1:N), X(K+1:N)) ) / A(K,K)
  ENDDO

END SUBROUTINE BACKSU

!meriki odigisi (partial pivoting)
SUBROUTINE pivot(A, B, k)
  IMPLICIT NONE

  DOUBLE PRECISION, INTENT (inout) :: A(:,:), B(:)
  INTEGER, INTENT (in) :: k
  INTEGER :: i, j, n
  DOUBLE PRECISION :: maxv

  INTERFACE
     ELEMENTAL SUBROUTINE swap(a,b)
       IMPLICIT NONE
       DOUBLE PRECISION, INTENT (inout) :: a, b
     END SUBROUTINE swap
  END INTERFACE

  n = SIZE(a,1)

  ! 'normalize' equations k and below
  DO i = k, n
     maxv = MAXVAL(ABS(A(i,:)))
     A(i,:) = A(i,:) / maxv
     B(i) = B(i) / maxv
  ENDDO

  ! find position of max element in k column below the diagonal.
  i = k 
  DO j = k+1, n
     IF ( ABS(a(i,k)) < ABS(a(j,k)) ) i = j
  ENDDO
  ! in i column a(i,k) is maximum.

  IF (i /= k) THEN              ! exchange k, i rows of A and B
     CALL swap(a(k,:), a(i,:))
     CALL swap(b(k), b(i))     
  END IF

END SUBROUTINE pivot

! a<->b
ELEMENTAL SUBROUTINE swap(a,b)
  IMPLICIT NONE
  DOUBLE PRECISION, INTENT (inout) :: a, b

  DOUBLE PRECISION :: temp

  temp = a
  a    = b
  b    = temp

END SUBROUTINE swap

